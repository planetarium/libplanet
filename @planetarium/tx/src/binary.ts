
export function compareUint8Array(a: Uint8Array, b: Uint8Array) {
  const length = Math.min(a.length, b.length);
  for (let i = 0; i < length; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}
