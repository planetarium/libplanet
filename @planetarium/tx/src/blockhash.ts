import { Encodable } from "bencodex";

export type BlockHash = Uint8Array;  // TODO: proper type definition

export function encodeBlockHash(blockHash: BlockHash): Encodable {
  if (blockHash.length !== 32) {
    throw new TypeError(`BlockHash must be 32 bytes long, but got ${
      blockHash.length} bytes.`);
  }
  return blockHash.buffer;
}
