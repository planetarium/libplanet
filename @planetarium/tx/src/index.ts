export { type Address, encodeAddress } from "./address";
export {
  type Currency,
  encodeCurrency,
  type FungibleAssetValue,
  getCurrencyHash,
  getMajorUnit,
  getMinorUnit,
  getSign,
} from "./assets";
export {
  encodeMint,
  encodeSystemAction,
  encodeTransfer,
  type Mint,
  type SystemAction,
  type Transfer,
} from "./action";
export type { TxMetadata, UnsignedTxWithSystemAction } from "./tx";
export { encodeTxMetadata, encodeUnsignedTxWithSystemAction } from "./tx";
