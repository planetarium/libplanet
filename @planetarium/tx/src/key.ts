import { Encodable } from "bencodex";

export type PublicKey = Uint8Array;  // TODO: proper type definition

export function encodePublicKey(publicKey: PublicKey): Encodable {
  if (publicKey.length < 1) {
    throw new TypeError("Public key must not be empty.");
  }
  return publicKey.buffer;
}
