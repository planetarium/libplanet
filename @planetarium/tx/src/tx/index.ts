export type { TxMetadata } from "./metadata";
export { encodeTxMetadata } from "./metadata";
export type { UnsignedTxWithSystemAction } from "./unsigned";
export { encodeUnsignedTxWithSystemAction } from "./unsigned";
