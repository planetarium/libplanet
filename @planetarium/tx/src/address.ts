import { Encodable } from "bencodex";
import { compareUint8Array } from "./binary";

export type Address = Uint8Array;  // TODO: proper type definition

export function encodeAddress(address: Address): Encodable {
  if (address.length !== 20) {
    throw new TypeError(`Address must be 20 bytes long, but got ${
      address.length} bytes.`);
  }
  return address.buffer;
}

export function encodeAddressSet(addresses: Set<Address>): Encodable {
  const array: Address[] = [];
  addresses.forEach(addr => array.push(addr));
  array.sort(compareUint8Array);
  return array.map(encodeAddress);
}
