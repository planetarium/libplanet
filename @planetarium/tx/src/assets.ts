import { Encodable, encode } from "bencodex";
import { Address, encodeAddressSet } from "./address";

export interface Currency {
  ticker: string;
  decimalPlaces: number;
  minters: Set<Address> | null;
  totalSupplyTrackable: boolean;
  maximumSupply: {
    major: bigint;
    minor: bigint;
  } | null;
}

export function encodeCurrency(currency: Currency): Encodable {
  const minters: Encodable = currency.minters === null
    ? null
    : encodeAddressSet(currency.minters);
  const serialized: Encodable = {
    ticker: currency.ticker,
    decimals: currency.decimalPlaces,
    minters,
  };

  if (currency.maximumSupply !== null) {
    if (!currency.totalSupplyTrackable) {
      throw new TypeError("maximumSupply implies totalSupplyTrackable");
    }

    serialized.maximumSupplyMajor = currency.maximumSupply.major;
    serialized.maximumSupplyMinor = currency.maximumSupply.minor;
  }

  if (currency.totalSupplyTrackable) {
    serialized.totalSupplyTrackable = true;
  }

  return serialized;
}

export async function getCurrencyHash(currency: Currency): Promise<Uint8Array> {
  const encoded = encode(encodeCurrency(currency));
  const buffer = await crypto.subtle.digest("SHA-1", encoded);
  return new Uint8Array(buffer);
}

export interface FungibleAssetValue {
  rawValue: bigint;
  currency: Currency;
}

function abs(value: bigint): bigint {
  return value < 0n ? -value : value;
}

export function getSign(value: FungibleAssetValue): -1 | 0 | 1 {
  return value.rawValue < 0n ? -1 : value.rawValue > 0n ? 1 : 0;
}

export function getMajorUnit(value: FungibleAssetValue): bigint {
  return abs(value.rawValue) / 10n ** BigInt(value.currency.decimalPlaces);
}

export function getMinorUnit(value: FungibleAssetValue): bigint {
  return abs(value.rawValue) % 10n ** BigInt(value.currency.decimalPlaces);
}
