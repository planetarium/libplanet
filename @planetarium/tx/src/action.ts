import { Encodable } from "bencodex";
import { Address, encodeAddress } from "./address";
import { encodeCurrency, FungibleAssetValue } from "./assets";

export type SystemAction = Mint | Transfer;

export function encodeSystemAction(action: SystemAction): Encodable {
  switch (action.type) {
  case "mint":
    return {
      type_id: 0,
      values: encodeMint(action),
    };
  case "transfer":
    return {
      type_id: 1,
      values: encodeTransfer(action),
    };
  default:
    throw new TypeError(`Unknown action type: ${action["type"]}`);
  }
}

export interface Mint {
  type: "mint";
  recipient: Address;
  amount: FungibleAssetValue;
}

export function encodeMint(action: Mint): Encodable {
  return {
    recipient: encodeAddress(action.recipient),
    currency: encodeCurrency(action.amount.currency),
    amount: action.amount.rawValue,
  };
}

export interface Transfer {
  type: "transfer";
  recipient: Address;
  amount: FungibleAssetValue;
}

export function encodeTransfer(action: Transfer): Encodable {
  return {
    recipient: encodeAddress(action.recipient),
    currency: encodeCurrency(action.amount.currency),
    amount: action.amount.rawValue,
  };
}
