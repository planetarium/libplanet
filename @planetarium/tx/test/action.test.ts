import { encode } from "bencodex";
import { expect, test } from "vitest";
import {
  type Mint,
  type Transfer,
  encodeMint,
  encodeTransfer,
  encodeSystemAction
} from "../src/action";
import { Address } from "../src/address";
import { Currency } from "../src/assets";
import { fromHex } from "./hex";

const FOO: Currency = {
  ticker: "FOO",
  decimalPlaces: 2,
  minters: null,
  totalSupplyTrackable: true,
  maximumSupply: null,
};
const addressA: Address = fromHex("D6D639DA5a58A78A564C2cD3DB55FA7CeBE244A9");

test("encodeSystemAction", () => {
  const mint: Mint = {
    type: "mint",
    recipient: addressA,
    amount: {
      rawValue: 12500n,
      currency: FOO,
    }
  };
  expect(encodeSystemAction(mint)).toEqual({
    type_id: 0,
    values: encodeMint(mint),
  });
  expect(encode(encodeSystemAction(mint))).toEqual(encode({
    type_id: 0,
    values: encodeMint(mint),
  }));
  const transfer: Transfer = {
    type: "transfer",
    recipient: addressA,
    amount: {
      rawValue: 12500n,
      currency: FOO,
    }
  };
  expect(encodeSystemAction(transfer)).toEqual({
    type_id: 1,
    values: encodeTransfer(transfer),
  });
  expect(encode(encodeSystemAction(transfer))).toEqual(encode({
    type_id: 1,
    values: encodeTransfer(transfer),
  }));
});

test("encodeMint", () => {
  const mint: Mint = {
    type: "mint",
    recipient: addressA,
    amount: {
      rawValue: 12500n,
      currency: FOO,
    }
  };
  const encoded = encodeMint(mint);
  expect(encoded).toMatchSnapshot();
  expect(encode(encoded)).toMatchSnapshot();
});

test("encodeTransfer", () => {
  const transfer: Transfer = {
    type: "transfer",
    recipient: addressA,
    amount: {
      rawValue: 12500n,
      currency: FOO,
    }
  };
  const encoded = encodeTransfer(transfer);
  expect(encoded).toMatchSnapshot();
  expect(encode(encoded)).toMatchSnapshot();
});
