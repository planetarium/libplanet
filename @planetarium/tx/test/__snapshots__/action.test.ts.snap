// Vitest Snapshot v1

exports[`encodeMint 1`] = `
{
  "amount": 12500n,
  "currency": {
    "decimals": 2,
    "minters": null,
    "ticker": "FOO",
    "totalSupplyTrackable": true,
  },
  "recipient": ArrayBuffer [],
}
`;

exports[`encodeMint 2`] = `
{
  "data": [
    100,
    117,
    54,
    58,
    97,
    109,
    111,
    117,
    110,
    116,
    105,
    49,
    50,
    53,
    48,
    48,
    101,
    117,
    56,
    58,
    99,
    117,
    114,
    114,
    101,
    110,
    99,
    121,
    100,
    117,
    56,
    58,
    100,
    101,
    99,
    105,
    109,
    97,
    108,
    115,
    105,
    50,
    101,
    117,
    55,
    58,
    109,
    105,
    110,
    116,
    101,
    114,
    115,
    110,
    117,
    54,
    58,
    116,
    105,
    99,
    107,
    101,
    114,
    117,
    51,
    58,
    70,
    79,
    79,
    117,
    50,
    48,
    58,
    116,
    111,
    116,
    97,
    108,
    83,
    117,
    112,
    112,
    108,
    121,
    84,
    114,
    97,
    99,
    107,
    97,
    98,
    108,
    101,
    116,
    101,
    117,
    57,
    58,
    114,
    101,
    99,
    105,
    112,
    105,
    101,
    110,
    116,
    50,
    48,
    58,
    214,
    214,
    57,
    218,
    90,
    88,
    167,
    138,
    86,
    76,
    44,
    211,
    219,
    85,
    250,
    124,
    235,
    226,
    68,
    169,
    101,
  ],
  "type": "Buffer",
}
`;

exports[`encodeTransfer 1`] = `
{
  "amount": 12500n,
  "currency": {
    "decimals": 2,
    "minters": null,
    "ticker": "FOO",
    "totalSupplyTrackable": true,
  },
  "recipient": ArrayBuffer [],
}
`;

exports[`encodeTransfer 2`] = `
{
  "data": [
    100,
    117,
    54,
    58,
    97,
    109,
    111,
    117,
    110,
    116,
    105,
    49,
    50,
    53,
    48,
    48,
    101,
    117,
    56,
    58,
    99,
    117,
    114,
    114,
    101,
    110,
    99,
    121,
    100,
    117,
    56,
    58,
    100,
    101,
    99,
    105,
    109,
    97,
    108,
    115,
    105,
    50,
    101,
    117,
    55,
    58,
    109,
    105,
    110,
    116,
    101,
    114,
    115,
    110,
    117,
    54,
    58,
    116,
    105,
    99,
    107,
    101,
    114,
    117,
    51,
    58,
    70,
    79,
    79,
    117,
    50,
    48,
    58,
    116,
    111,
    116,
    97,
    108,
    83,
    117,
    112,
    112,
    108,
    121,
    84,
    114,
    97,
    99,
    107,
    97,
    98,
    108,
    101,
    116,
    101,
    117,
    57,
    58,
    114,
    101,
    99,
    105,
    112,
    105,
    101,
    110,
    116,
    50,
    48,
    58,
    214,
    214,
    57,
    218,
    90,
    88,
    167,
    138,
    86,
    76,
    44,
    211,
    219,
    85,
    250,
    124,
    235,
    226,
    68,
    169,
    101,
  ],
  "type": "Buffer",
}
`;
